File: Java/Longest Substring with At Most Two Distinct Characters.java
/*
Given a string S, find the length of the longest substring T that contains at most two distinct characters.
For example,
Given S = “eceba”,
T is "ece" which its length is 3.
*/

//O(n^2)
public int lengthOfLongestSubstringTwoDistinct(String s) {
   int start = 0, end = -1, maxLen = 0;
   for (int k = 1; k < s.length(); k++) 
   {
      if (s.charAt(k) == s.charAt(k - 1)) continue;
      if (end >= 0 && s.charAt(end) != s.charAt(k)) 
      {
      	maxLen = Math.max(k - start, maxLen);
		start = end + 1; 
	}
		end = k - 1; 
	}
   return Math.max(s.length() - start, maxLen);
}

//O(n^2), with hash table
public int lengthOfLongestSubstringTwoDistinct(String s) {
   int[] count = new int[256];
   int i = 0, numDistinct = 0, maxLen = 0;
   for (int j = 0; j < s.length(); j++) {
      if (count[s.charAt(j)] == 0) numDistinct++;
      count[s.charAt(j)]++;
      while (numDistinct > 2) {
         count[s.charAt(i)]--;
         if (count[s.charAt(i)] == 0) numDistinct--;
         i++;
     }
      maxLen = Math.max(j - i + 1, maxLen);
   }
   return maxLen;
}

File: Python/Longest Substring with At Most Two Distinct Characters.py
"""
Given a string, find the length of the longest substring T that contains at most 2 distinct characters.

For example, Given s = “eceba”,

T is "ece" which its length is 3.
"""

class Solution(object):
    def lengthOfLongestSubstringTwoDistinct(self, s):
        """
        :type s: str
        :rtype: int
        """
        longest, start, distinct_count, visited = 0, 0, 0, [0 for _ in xrange(256)]
        
        for i, char in enumerate(s):
            if visited[ord(char)] == 0:
                distinct_count += 1
            visited[ord(char)] += 1
            
            while distinct_count > 2:
                visited[ord(s[start])] -= 1
                if visited[ord(s[start])] == 0:
                    distinct_count -= 1
                start += 1
                
            longest = max(longest, i - start + 1)
        return longest
        